v1alpha2 APIs:

jacobweinstock opened the pull request:
## Description

<!--- Please describe what this PR is going to change -->
Work in progress for the v1alpha2 API spec. There are example yamls in `docs/technical/v1alpha2` These are breaking API changes. Not only do the specs change drastically, but some of the names of the CRDs have been changed.

| v1alpha1 | v1alpha2 | name change |
| :-------- | :--------- | :--------- |
| hardware.tinkerbell.org | hardware.tinkerbell.org | no |
| template.tinkerbell.org | task.tinkerbell.org | yes |
| workflow.tinkerbell.org | workflow.tinkerbell.org | no |
| workflowruleset.tinkerbell.org | policy.tinkerbell.org | yes |
| machine.bmc.tinkerbell.org | bmc.tinkerbell.org | yes |
| job.bmc.tinkerbell.org | job.bmc.tinkerbell.org | no |
| task.bmc.tinkerbell.org | - | will no longer exist |

samples:

[hardware](https://github.com/jacobweinstock/tinkerbell/blob/v1alpha2-implementation/docs/technical/v1alpha2/hardware.yaml)
[task](https://github.com/jacobweinstock/tinkerbell/blob/v1alpha2-implementation/docs/technical/v1alpha2/task.yaml)
[workflow](https://github.com/jacobweinstock/tinkerbell/blob/v1alpha2-implementation/docs/technical/v1alpha2/workflow.yaml)
[bmc](https://github.com/jacobweinstock/tinkerbell/blob/v1alpha2-implementation/docs/technical/v1alpha2/bmc.yaml)
[job](https://github.com/jacobweinstock/tinkerbell/blob/v1alpha2-implementation/docs/technical/v1alpha2/job.yaml)
[policy](https://github.com/jacobweinstock/tinkerbell/blob/v1alpha2-implementation/docs/technical/v1alpha2/policy.yaml)

The Hardware object also changes to use the NoCloud cloud-init datasource runtime configurations. These configs, and implementation in Tootles, will provide a lot more flexible Cloud-init support than the current Tootles default Amazon EC2 datasource. All of the following fields, `hardware.spec.instance.{metadata,userdata,vendordata,networkConfig}`, support templating. This means that you can reference other parts of the Hardware object. The rendering of these fields will happen via Tootles (when the http(s) endpoints for each are hit) and via the Tink controller (when the fields are referenced in a Workflow and Tink controller renders the Workflow in a Pipeline). 

The Hardware object now has full support for IPv6.

Fixes: #

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->


## How are existing users impacted? What migration steps/scripts do we need?

<!--- Fixes a bug, unblocks installation, removes a component of the stack, etc -->
<!--- Requires a DB migration script, etc. -->


## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

jacobweinstock committed:
Add v1alpha2 API types and CRDs:

- Add v1alpha2 Task CRD (reusable action definitions)
- Add v1alpha2 Workflow CRD (execution orchestrator)
- Add v1alpha2 Hardware, BMC, and Job CRDs
- Add v1alpha2 Policy CRD type
- Add generated deepcopy and CRD manifests
- Update api/ module dependencies

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add v1alpha2 API types and CRDs:

- Add v1alpha2 Task CRD (reusable action definitions)
- Add v1alpha2 Workflow CRD (execution orchestrator)
- Add v1alpha2 Hardware, BMC, and Job CRDs
- Add v1alpha2 Policy CRD type
- Add generated deepcopy and CRD manifests
- Update api/ module dependencies

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add v1alpha2 Policy CRD:

- Add generated Policy CRD manifest for v1alpha2

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add v1alpha2 Policy CRD:

- Add generated Policy CRD manifest for v1alpha2

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Refactor CRD loading to use embed.FS with versioned directories:

- Move v1alpha1 CRD files into crd/bases/v1alpha1/
- Update crd.go to load CRDs by version (v1alpha1, v1alpha2)
- Add CRD loading tests

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Refactor CRD loading to use embed.FS with versioned directories:

- Move v1alpha1 CRD files into crd/bases/v1alpha1/
- Update crd.go to load CRDs by version (v1alpha1, v1alpha2)
- Add CRD loading tests

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add v1alpha2 CRD support to dashboard UI:

- Add kind-to-route mappings for v1alpha2 types (Task, Workflow, BMC, Policy)
- Add v1alpha2 kind descriptions
- Update dashboard to support multiple API versions
- Add UI tests for v1alpha2 CRD metadata

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add v1alpha2 CRD support to dashboard UI:

- Add kind-to-route mappings for v1alpha2 types (Task, Workflow, BMC, Policy)
- Add v1alpha2 kind descriptions
- Update dashboard to support multiple API versions
- Add UI tests for v1alpha2 CRD metadata

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add v1alpha2 example documentation:

- Add example YAML for Task (reusable action definitions)
- Add example YAML for Workflow (execution orchestrator)
- Add example YAML for Policy (auto-creation rules)
- Add example YAML for Hardware, BMC, and Job resources

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add v1alpha2 example documentation:

- Add example YAML for Task (reusable action definitions)
- Add example YAML for Workflow (execution orchestrator)
- Add example YAML for Policy (auto-creation rules)
- Add example YAML for Hardware, BMC, and Job resources

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Build and tooling improvements:

- Add stamp-file-based incremental builds for code generation
- Add CRD manifest generation targets for v1alpha1 and v1alpha2
- Regenerate stringer files
- Update .gitignore for build outputs

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Build and tooling improvements:

- Add stamp-file-based incremental builds for code generation
- Add CRD manifest generation targets for v1alpha1 and v1alpha2
- Regenerate stringer files
- Update .gitignore for build outputs

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

brk0v commented on api/v1alpha2/tinkerbell/hardware.go:
```diff
@@ -0,0 +1,431 @@
+	// Router is the IP address of the router for this route.
+	// +kubebuilder:validation:Pattern=`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$`
+	Router string `json:"router"`
+}
```
Adding an optional `Interface` field here could simplify the future IPv6 support work.

jacobweinstock replied:
Ah yes. We're going to need a lot of new fields for IPv6. I will update the spec. Thanks!

jacobweinstock replied:
Hey @brk0v. I've updated the hardware spec with IPv6 fields. Let me know what you think. Thanks!

brk0v commented on docs/technical/v1alpha2/hardware.yaml:
```diff
@@ -0,0 +1,121 @@
+          eth0:
+            addresses:
+              - {{ (index .hardware.networkInterfaces "52:54:00:41:05:c6").ipam.ipv4.address }}/{{ (index .hardware.networkInterfaces "52:54:00:41:05:c6").ipam.ipv4.prefix }}
+            gateway4: {{ (index .hardware.networkInterfaces "52:54:00:41:05:c6").ipam.ipv4.gateway }}
```
nit: `netplan` deprecated `gateway4` and `gateway6` fields. If it's a `netplan` config, probable using the `routes` section would be a better example.

jacobweinstock replied:
updated, thanks for catching that!

jacobweinstock committed:
Add DHCPv6 and restructure DHCP in v1alpha2 Hardware spec:

DHCPv6 support is needed for IPv6 network booting and configuration.
The flat DHCPv4/DHCPv6 fields on NetworkInterface are restructured
under a DHCP parent (dhcp.v4/v6) to mirror the existing IPAM pattern
(ipam.ipv4/ipv6).

This will support both SLAAC (Stateless Address Autoconfiguration)
and DHCPv6. The implementation of this will be
client-driven: clients send either Information-Request or Solicit
and the server responds based on the message type.

Timeservers is renamed to ntpServers and domainSearchList is added
to DHCPv4 for naming consistency and option parity with DHCPv6.

Nameserver and Timeserver validation patterns are extended to accept
IPv6 addresses.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add DHCPv6 and restructure DHCP in v1alpha2 Hardware spec:

DHCPv6 support is needed for IPv6 network booting and configuration.
The flat DHCPv4/DHCPv6 fields on NetworkInterface are restructured
under a DHCP parent (dhcp.v4/v6) to mirror the existing IPAM pattern
(ipam.ipv4/ipv6).

This will support both SLAAC (Stateless Address Autoconfiguration)
and DHCPv6. The implementation of this will be
client-driven: clients send either Information-Request or Solicit
and the server responds based on the message type.

Timeservers is renamed to ntpServers and domainSearchList is added
to DHCPv4 for naming consistency and option parity with DHCPv6.

Nameserver and Timeserver validation patterns are extended to accept
IPv6 addresses.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add lookupID field to Hardware v1alpha2 Instance spec:

When users bring their own DHCP, Tootles cannot use source IP to map
metadata requests to Hardware objects. The lookupID field provides
an identifier that Tootles can index and match against the NoCloud
seedfrom URL path to route requests to the correct Hardware.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add lookupID field to Hardware v1alpha2 Instance spec:

When users bring their own DHCP, Tootles cannot use source IP to map
metadata requests to Hardware objects. The lookupID field provides
an identifier that Tootles can index and match against the NoCloud
seedfrom URL path to route requests to the correct Hardware.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Refine v1alpha2 API specs:

- Add IPXE script/url mutual exclusivity validation
- Require at least one of hardwareRef or bmcRef in WorkflowHardware
- Rename WorkflowConfig.Extra to Globals for clarity
- Support multiple Tasks per Policy via PolicyWorkflowTask
- Remove TaskSpec.Name (use metadata.name instead)
- Fix Action command pattern and add Retries minimum validation
- Make BMC Connection fields optional where appropriate
- Add Background field to Action for non-blocking execution (kexec, reboot, etc.)
- Fix typos in status comments
- Regenerate CRDs

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Refine v1alpha2 API specs:

- Add IPXE script/url mutual exclusivity validation
- Require at least one of hardwareRef or bmcRef in WorkflowHardware
- Rename WorkflowConfig.Extra to Globals for clarity
- Support multiple Tasks per Policy via PolicyWorkflowTask
- Remove TaskSpec.Name (use metadata.name instead)
- Fix Action command pattern and add Retries minimum validation
- Make BMC Connection fields optional where appropriate
- Add Background field to Action for non-blocking execution (kexec, reboot, etc.)
- Fix typos in status comments
- Regenerate CRDs

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>

jacobweinstock committed:
Add raspberry pi and pxelinux fields:

This supports the custom fields need to get RPi's
and other SBC's that work with uboot working.

The other change was just a bump of the controller-gen
cli command. Generated files are only updated with
a code comment to the new version. No structural changes.

Signed-off-by: Jacob Weinstock <jakobweinstock@gmail.com>
